image-rs/src/verity.rs

```diff
@@ -0,0 +1,144 @@
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct DmVerityOption {
+    /// Hash algorithm for dm-verity.
+    pub hashtype: String,
+    /// Used block size for the data device.
+    pub blocksize: String,
+    /// Used block size for the hash device.
+    pub hashsize: String,
+    /// Size of data device used in verification.
+    pub blocknum: String,
+    /// Offset of hash area/superblock on hash_device.
+    pub offset: String,
+    /// Root hash for device verification or activation.
+    pub hash: String,
+}
+
+/// Creates a mapping with <name> backed by data_device <source_device_path>
+/// and using hash_device for in-kernel verification.
+/// It will return the verity block device Path "/dev/mapper/<name>"
+/// Notes: the data device and the hash device are the same one.
+pub fn create_verity_device(
+    verity_option: &DmVerityOption,
+    source_device_path: &str,
+) -> Result<String> {
+    let veritysetup_path = veritysetup_exists()?;
+
+    let output = Command::new(veritysetup_path)
+        .args(&[
+            "open",
+            "--no-superblock",
+            "--format=1",
+            "-s",
+            "",
+            &format!("--hash={}", verity_option.hashtype),
+            &format!("--data-block-size={}", verity_option.blocksize),
+            &format!("--hash-block-size={}", verity_option.hashsize),
+            "--data-blocks",
+            verity_option.blocknum.as_str(),
+            "--hash-offset",
+            verity_option.offset.as_str(),
+            source_device_path,
+            verity_option.hash.as_str(),
+            source_device_path,
+            verity_option.hash.as_str(),
+        ])
+        .output()?;
+    if output.status.success() {
+        return Ok(format!("{}{}", "/dev/mapper/", verity_option.hash.as_str()));
+    } else {
+        let error_message = String::from_utf8_lossy(&output.stderr);
+        bail!("Failed to create dm-verity device: {}", error_message);
+    }
+}
+pub fn decode_verity_options(verity_options: &str) -> Result<DmVerityOption> {
+    let decoded = base64::decode(verity_options)?;
+    let parsed_data = serde_json::from_slice::<DmVerityOption>(&decoded)?;
+    Ok(parsed_data)
+}
+
+pub fn attach_data_to_loop_device(data_path: &str) -> Result<LoopDevice> {
+    let lc = loopdev::LoopControl::open()?;
+    let ld = lc.next_free()?;
+    ld.attach(data_path, 0)?;
+    Ok(ld)
+}
+pub fn get_verity_device_name(verity_options: &str)-> Result<String>{
+    let decoded_data = decode_verity_options(verity_options)?;
+    Ok(decoded_data.hash)
+}
+/// Calculates and permanently stores hash verification data for data_device.
+/// It will return the root hash.
+pub fn create_hash_device(
+    verity_options: &mut DmVerityOption,
+    data_device_path: &str,
+) -> Result<String> {
+    let veritysetup_path: &str = veritysetup_exists()?;
+    let output = Command::new(veritysetup_path)
```

Comment:
How about using https://docs.rs/devicemapper/latest/devicemapper/struct.DM.html to get rid of the dependency on veritysetup binary?

Reply:
> I haven't used devicemapper, but @jiangliu's comment seems right to me. Related to this: it looks like `verity_option` doesn't get sanity-checked or sanitized anywhere, so its values get plugged directly into the args to `versitysetup` when issuing the `Command`. Using a library call (@jiangliu's suggestion) and/or sanity-checking those values might make error handling/checking easier.

Yes, I agree with @jiangliu's comment and I am working on using the devicemapper library to handle/check the options.